core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java

```diff
@@ -109,50 +112,58 @@ private void setPrimariesInitialRecoveries(int primariesInitialRecoveries) {
 
     @Override
     public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
-        if (shardRouting.primary()) {
-            assert shardRouting.unassigned() || shardRouting.active();
-            if (shardRouting.unassigned()) {
-                // primary is unassigned, means we are going to do recovery from gateway
-                // count *just the primary* currently doing recovery on the node and check against concurrent_recoveries
-                int primariesInRecovery = 0;
-                for (ShardRouting shard : node) {
-                    // when a primary shard is INITIALIZING, it can be because of *initial recovery* or *relocation from another node*
-                    // we only count initial recoveries here, so we need to make sure that relocating node is null
-                    if (shard.initializing() && shard.primary() && shard.relocatingNodeId() == null) {
-                        primariesInRecovery++;
-                    }
+        assert shardRouting.primary() == false || shardRouting.unassigned() || shardRouting.active();
+
+        if (shardRouting.primary() == false || shardRouting.unassigned() == false) {
```

Comment:
so this means that we either have a replica peer recovery (primary == false) or we have a primary relocation that means it's not assigned? I wonder if we can make this more clear? maybe assign local vars?